velox/functions/prestosql/aggregates/benchmarks/ReduceAgg.cpp

```diff
@@ -0,0 +1,232 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <folly/Benchmark.h>
+#include <folly/init/Init.h>
+#include <string>
+
+#include "velox/exec/tests/utils/Cursor.h"
+#include "velox/exec/tests/utils/HiveConnectorTestBase.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+#include "velox/vector/fuzzer/VectorFuzzer.h"
+
+DEFINE_int64(fuzzer_seed, 99887766, "Seed for random input dataset generator");
+
+using namespace facebook::velox;
+using namespace facebook::velox::connector::hive;
+using namespace facebook::velox::exec::test;
+
+static constexpr int32_t kNumVectors = 10;
+static constexpr int32_t kRowsPerVector = 10'000;
+
+namespace {
+
+// Compare performance of sum(x) with equivalent reduce_agg(x,..).
+class ReduceAggBenchmark : public HiveConnectorTestBase {
+ public:
+  explicit ReduceAggBenchmark() {
+    OperatorTestBase::SetUpTestCase();
+    HiveConnectorTestBase::SetUp();
+
+    inputType_ = ROW({
+        {"i16", SMALLINT()},
+    });
+
+    VectorFuzzer::Options opts;
+    opts.vectorSize = kRowsPerVector;
+    opts.nullRatio = 0.1;
+    VectorFuzzer fuzzer(opts, pool(), FLAGS_fuzzer_seed);
+
+    std::vector<RowVectorPtr> vectors;
+    for (auto i = 0; i < kNumVectors; ++i) {
+      vectors.emplace_back(fuzzer.fuzzInputRow(inputType_));
+    }
+
+    filePath_ = TempFilePath::create();
+    writeToFile(filePath_->path, vectors);
+  }
+
+  ~ReduceAggBenchmark() override {
+    HiveConnectorTestBase::TearDown();
+  }
+
+  void TestBody() override {}
+
+  static inline const std::string kSum = "sum(i16)";
+  static inline const std::string kReduceAgg =
+      "reduce_agg(i16, 0, (x, y) -> (x + y), (x, y) -> (x + y))";
+
+  void verify() {
+    // Verify global aggregation.
+    {
+      auto sumResults = runOnce(makeGlobalPlan(kSum));
+      auto reduceAggResults = runOnce(makeGlobalPlan(kReduceAgg));
+
+      assertEqualResults({sumResults}, {reduceAggResults});
+    }
+
+    // Verify group by aggregation.
+    {
+      auto sumResults = runOnce(makeGroupByPlan(kSum));
+      auto reduceAggResults = runOnce(makeGroupByPlan(kReduceAgg));
+
+      assertEqualResults({sumResults}, {reduceAggResults});
+    }
+  }
+
+  void runGlobalSum() {
+    runGlobal(kSum);
+  }
+
+  void runGroupBySum() {
+    runGlobal(kSum);
+  }
+
+  void runGlobalReduceAgg() {
+    runGlobal(kReduceAgg);
+  }
+
+  void runGroupByReduceAgg() {
+    runGroupBy(kReduceAgg);
+  }
+
+ private:
+  RowVectorPtr runOnce(const core::PlanFragment& planFragment) {
```

Comment:
maybe we could move this function to the test base class? Seems pretty general and useful

Reply:
The base class I use here is used for unit tests and in that context these functions are not very useful. Maybe we could create a base class for benchmarks specifically similar to what Laith created for benchmarking expressions.